uni/lib/model/providers/plausible/plausible_provider.dart

```diff
@@ -0,0 +1,116 @@
+import 'package:logger/logger.dart';
+import 'package:plausible_analytics/plausible_analytics.dart';
+import 'package:provider/provider.dart';
+import 'package:uni/controller/local_storage/preferences_controller.dart';
+
+class PlausibleProvider extends StatefulWidget {
+  const PlausibleProvider({
+    required this.plausible,
+    required this.child,
+    super.key,
+  });
+
+  final Plausible? plausible;
+  final Widget child;
+
+  @override
+  State<StatefulWidget> createState() => _PlausibleProviderState();
+}
+
+class _PlausibleProviderState extends State<PlausibleProvider> {
+  int _batteryLevel = 0;
+  bool _isInBatterySaveMode = true;
+  ConnectivityResult _connectivityResult = ConnectivityResult.mobile;
+  bool _isUsageStatsEnabled = true;
+
+  bool _canUpdateBatteryState = true;
+
+  @override
+  void initState() {
+    super.initState();
+
+    final plausible = widget.plausible;
+    if (plausible != null) {
+      plausible.enabled = false;
+
+      _startListeners(plausible)
+          .then((_) => _updateBatteryState())
+          .then((_) => _updateConnectivityState())
+          .then((_) => _updateUsageStatsState());
+    }
+  }
+
+  void _updateAnalyticsState() {
+    final plausible = widget.plausible;
+    if (plausible != null) {
+      plausible.enabled = _isUsageStatsEnabled &&
+          !_isInBatterySaveMode &&
+          _batteryLevel > 20 &&
+          _connectivityResult == ConnectivityResult.wifi;
+
+      Logger().d('Plausible enabled: ${plausible.enabled}');
+    }
+  }
+
+  Future<void> _updateBatteryState() async {
+    if (!_canUpdateBatteryState) {
+      return;
+    }
+
+    _canUpdateBatteryState = false;
+    Timer(const Duration(seconds: 10), () => _canUpdateBatteryState = true);
+
+    final battery = Battery();
+    _batteryLevel = await battery.batteryLevel;
+    _isInBatterySaveMode = await battery.isInBatterySaveMode;
+
+    _updateAnalyticsState();
+  }
+
+  Future<void> _updateConnectivityState() async {
+    final connectivity = Connectivity();
+    _connectivityResult = await connectivity.checkConnectivity();
+
+    _updateAnalyticsState();
+  }
+
+  Future<void> _updateUsageStatsState() async {
+    _isUsageStatsEnabled = PreferencesController.getUsageStatsToggle();
+    _updateAnalyticsState();
+  }
+
+  Future<void> _startListeners(Plausible plausible) async {
+    final connectivity = Connectivity();
+    connectivity.onConnectivityChanged.listen((result) {
+      _connectivityResult = result;
+      _updateAnalyticsState();
+    });
+
+    PreferencesController.onStatsToggle.listen((event) {
+      _isUsageStatsEnabled = event;
+      _updateAnalyticsState();
+    });
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Listener(
+      behavior: HitTestBehavior.translucent,
+      onPointerDown: (event) {
+        final plausible = widget.plausible;
+        if (plausible != null) {
+          _updateBatteryState();
+        }
+      },
```

Comment:
So, if I am understanding this correctly, this adds up to the Listener automatically injected by `Plausible`. Where is the logic to disable the plausible event handler when the battery is low?

Reply:
This function calls updateBatteryState, which then calls updateAnalyticsState to disable the Plausible instance. The Plausible event listener for the navigation is still there, but when it calls plausible.event to trigger an event, the call does nothing (no request is made) because the Plausible instance is disabled.

Reply:
Plausible doesn't inject a Listener per se. The Plausible object is just a wrapper around a network request that is made to the plausible instance. It can be enabled or disabled to allow/stop network requests.

The NavigationObserver remains working even when the Plausible instance is disabled, but all events that are registered by it are ignored if the Plausible instance is disabled. (Also, the navigation observer doesn't listen to clicks, it listens to navigation changes)

Reply:
Ok, thank you!